metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationDriver.java

```diff
@@ -619,34 +679,45 @@ public void run() throws Exception {
         }
     }
 
-    class BecomeZkControllerEvent extends MigrationEvent {
+    class BecomeZkControllerEvent extends MigrationEvent implements MigrationWriteEvent {
+        private final LeaderAndEpoch leaderAndEpoch;
+
+        BecomeZkControllerEvent(LeaderAndEpoch leaderAndEpoch) {
+            this.leaderAndEpoch = leaderAndEpoch;
+        }
+
         @Override
         public void run() throws Exception {
+            // The leader epoch check in checkDriverState prevents us from getting stuck retrying this event after a
+            // new leader has been seen.
             if (checkDriverState(MigrationDriverState.BECOME_CONTROLLER, this)) {
-                applyMigrationOperation("Claiming ZK controller leadership", zkMigrationClient::claimControllerLeadership);
+                applyMigrationOperation("Claimed ZK controller leadership", zkMigrationClient::claimControllerLeadership, true);
                 if (migrationLeadershipState.zkControllerEpochZkVersion() == ZkMigrationLeadershipState.UNKNOWN_ZK_VERSION) {
                     log.info("Unable to claim leadership, will retry until we learn of a different KRaft leader");
+                    return; // Stay in BECOME_CONTROLLER state and retry
+                }
+
+                // KAFKA-16171 and KAFKA-16667: Prior writing to /controller and /controller_epoch ZNodes above,
+                // the previous controller could have modified the /migration ZNode. Since ZK does grant us linearizability
+                // between writes and reads on different ZNodes, we need to write something to the /migration ZNode to
+                // ensure we have the latest /migration zkVersion.
+                applyMigrationOperation("Updated migration state", state -> {
+                    state = state.withMigrationZkVersion(-1); // Causes an unconditional update on /migration
```

Comment:
Suggestion, as it's not obvious how this leads to the unconditional update.

```suggestion
                    state = state.withMigrationZkVersion(-1); // Causes an unconditional update on /migration via KafkaZkClient#retryRequestsUntilConnected
```